k9mail/src/main/java/com/fsck/k9/ui/crypto/MessageCryptoHelper.java

```diff
@@ -579,8 +583,17 @@ private void addCryptoResultAnnotationToMessage(CryptoResultAnnotation resultAnn
     }
 
     private void onCryptoFinished() {
-        currentCryptoPart = null;
-        partsToDecryptOrVerify.removeFirst();
+        boolean currentPartIsFirstInQueue = partsToDecryptOrVerify.peekFirst() == currentCryptoPart;
+        if (!currentPartIsFirstInQueue) {
+            throw new IllegalStateException(
+                    "Trying to remove part from queue that is not the currently processed one!");
+        }
+        if (currentCryptoPart != null) {
```

Comment:
Shouldn't this check be first.

If `partsToDecryptOrVerify` is empty. `peekFirst` will return null. If `currentCryptoPart` is also null we will throw. Whereas we should be logging an error.

I'm wondering if this scenario happens when the PgpApi returns two results for one part - maybe an error and a timeout response?